core/src/main/scala-2/zio/config/TupleConversion.scala

```diff
@@ -0,0 +1,86 @@
+package zio.config
+
+import scala.language.experimental.macros
+import scala.reflect.macros.whitebox
+
+trait TupleConversion[A, B] {
+  def to(a: A): B
+  def from(b: B): A
+}
+object TupleConversion {
+  def apply[Prod, Tup]: TupleConversion[Prod, Tup] = macro genTupleConversion[Prod, Tup]
+
+  def genTupleConversion[Prod: c.WeakTypeTag, Tup: c.WeakTypeTag](
+    c: whitebox.Context
+  ): c.Expr[TupleConversion[Prod, Tup]] = {
+    import c.universe._
+    val prodTpe = c.weakTypeOf[Prod]
+    if (!prodTpe.typeSymbol.isClass ||
+        !prodTpe.typeSymbol.asClass.isCaseClass) {
+      c.abort(c.enclosingPosition, s"Type ${prodTpe.typeSymbol} is not a case class")
+    }
+    val paramLists = prodTpe.typeSymbol.asClass.primaryConstructor.asMethod.typeSignatureIn(prodTpe).paramLists
+    val result = paramLists match {
+      case List(List(singleParam)) =>
+        // Special case: single parameter products
+
+        val typ = singleParam.typeSignatureIn(prodTpe).finalResultType
+
+        q"""new _root_.zio.config.TupleConversion[$prodTpe, $typ] {
+                override def to(a: $prodTpe): $typ = a.${singleParam.name.toTermName}                
+                override def from(b: $typ): $prodTpe =
+                    new ${prodTpe}(b)
+        }
+        """
```

Comment:
Couldn't `case List(params)` handle single parameter products as well?

Reply:
For n>1 we assume that `Tup` is a tuple and deconstruct it to the case class parameters. For n=1 we assume its not a tuple and use the value directly. This makes it more usable in practice as you can write `string("foo").to[X]` for example. 
In the Scala3 version there is an additional implicit that automatically converts `A` to `Tuple1[A]` first.

Reply:
Ah. Ofcourse. Makes sense.